### ISSUE TITLE
hour input should be hours, not time stamp

### ISSUE THREAD

**Issue by flah00** [bushwickfc/bfc-members#16]

**Comment by flah00** [comment]
1. Remove `start` and `finish` from views members/time_banks and time_banks
  1. Update `TimeBanksController` and `Member::TimeBanksController`, remove `start` and `finish` from `#time_bank_params`
1. Add new attributes to `TimeBank`, `date_worked`, `hours_worked`
  1. Add `date_worked` and `hours_worked` to `TimeBanksController` and `Member::TimeBanksController` `#time_bank_params` 
  1. when `#hours_worked` is updated, it should update `finish`
  1. Add new attributes to views members/time_banks and time_banks
1. Change datetime to date picker
1. Add tests, to verify changes

**Comment by stratigos** [comment]
Hello,

For the existing timebanks controller tests, should I go ahead and swap out the start/finish params, and replace with date_worked, hours_worked (and appropriate values)?

**Comment by stratigos** [comment]
Also I have found the time_banks for committees as well. Ill go ahead and change these too.

**Comment by stratigos** [comment]
It also looks like app/assets/javascripts/time_banks.js wont be needed anymore. Let me know if I should axe that as well, Ill leave it there for now.

**Comment by stratigos** [comment]
The update I just made to include datetimepicker was a bit tricky for someone newish to Rails, and I think Im doing something unconventional. Please let me know if there is a better way to handle the following situation.

The datetimepicker requires its own JS asset and a dependency (moment.js). These have to be loaded after bootstrap.js. The bootstrap.js file is hard coded into the application layout, and appears after the application asset manifest call. Thus, placing `bootstrap-datetimepicker` and `moment` assets wont work if placed anywhere in the manifest file (app/assets/javascripts/application.js). 

I got around that by creating initializers/assets.rb, and precompiling these two assets, and then calling them in application layout after bootstrap.js is requested.

The next problem, however, is with loading the javascript required to transform the input field in the web form itself into a datetimepicker. I had initially placed this call ( `$('#elementID).datetimepicker();` ) inside app/assets/javascripts/time_banks.js, though this also runs into the sort-of-race-condition I described above. Namely, on initial page load, time_banks.js loads before bootstrap.js and before the `bootstrap-datetimepicker` asset(s), and thus does nothing when clicked on.

So I placed this javascript call to `datetimepicker();` within the erb views themselves. This not only duplicated code, but seems like it mixes concerns, and breaks Rails' JS asset pipeline conventions. I did try to create a javascript file outside of the asset manifest, but it does not ignore subdirectories as stated in the docs, and I cant find any way of expressing a file/subdir exclude option.

For a greater solution, my first intuition is to remove TWBS/bootstrap.js from the application layout, and load it with a bootstrap rubygem and place it within the application asset manifest (application.js). This would allow jQuery, TWBS, bootstrap-datetimepicker, and the time_banks.js files to load in an expected order, and follow the asset pipeline conventions. 

There could easily be a better or 'more-Railsy' way of doing this, I just havent had the experience to see the right trick.

Could anyone suggest a more conventional, concern-separated, DRY way of doing the above?

**Comment by themikehogg** [comment]
Wow,
What's wrong with just using the rails date picker?

**Comment by stratigos** [comment]
@themikehogg Allow me to reiterate once more that I am relatively new to the Rails framework. When I google search for "Rails datepicker" or "Rails datetimepicker" all I get are results for various TWBS datetimepicker gems. I had assumed this was what was being requested in step 3 of the [instructions above](https://github.com/bushwickfc/bfc-members/issues/16#issuecomment-111843584).

Also the default Rails date picker seemed cumbersome for UX, since there are several select options to manipulate, and the bootstrap datetimepicker is essentially just one button click (and looks nicer). So yah I just made the wrong assumption and overcomplicated step 3. However, the issue of the sequence of JS dependencies loading will still persist, and future revisions which depend on bootstrap.js will likely run into the same problem.

Ill ditch the last commit, and stash it somewhere locally, and just redo this with the Rails default. Though, if only date picker is used, and not datetime picker, I think the whole system involving TimeBank.start may need to be addressed, since the various calculations involving that parameter will become inaccurate. I may not understand fully how TimeBank data is calculated, but it seems to me like either TimeBank.start needs a value accurate to the minute, or, TimeBank.date_worked needs to fully replace TimeBank.start.

**Comment by stratigos** [comment]
Ok refs 271a83c485b932c7b1c8540d2c53cb866d15310d and a3348c185d4eab48d0571c2c545978aa197a1514 are removed, and branch 16_hourInput has been rebased into 079b98fcbecd7ca14cfc19c16c8a61497b4abf02 (just removes the bootstrap-datetimepicker stuff).

**Comment by themikehogg** [comment]
Sorry I didn't mean to upset the apple cart. I have been there myself, where a simple approach gradually becomes more of an effort than it seems it should be.  Also, I am just trying to help.  You are the one working on it, who has the time, so thank you :)

I agree with you. I was going to also look for a better date control than the rails one.  My approach was to get the page working with the rails date control, 3 drop downs, and then use a jquery datepicker.  it sounds like I might have run into your difficulty eventually.  

Reading your 8/1 message, step by step, makes sense to me, butat this part:
> So I placed this javascript call to `datetimepicker();` within the erb views themselves. This not only duplicated code, bu

Why not put your datetimepicker() inside the document.ready().  That's a pretty standard way to do it and sounds like what you want here, right?

I agree that it sounds like it should be removed from application.html.erb and placed in assets.  It's also confusing that it's a gem.  I would remove it from both of these places, and download the files into the assets directories, based on some answers I see here:

http://stackoverflow.com/questions/18371318/installing-bootstrap-3-on-rails-app

Hope that helps.  Thanks again!
Mike


On Sun, 02 Aug 2015 09:33:37 -0700
Todd <notifications@github.com> wrote:

> @themikehogg Allow me to reiterate once more that I am relatively new to the Rails framework. When I google search for "Rails datepicker" or "Rails datetimepicker" all I get are results for various TWBS datetimepicker gems. I had assumed this was what was being requested in step 3 of the [instructions above](https://github.com/bushwickfc/bfc-members/issues/16#issuecomment-111843584).
> 
> Also the default Rails date picker seemed cumbersome for UX, since there are several select options to manipulate, and the bootstrap datetimepicker is essentially just one button click (and looks nicer). So yah I just made the wrong assumption and overcomplicated step 3. However, the issue of the sequence of JS dependencies loading will still persist, and future revisions which depend on bootstrap.js will likely run into the same problem.
> 
> Ill ditch the last commit, and stash it somewhere locally, and just redo this with the Rails default. Though, if only date picker is used, and not datetime picker, I think the whole system involving TimeBank.start may need to be addressed, since the various calculations involving that parameter will become inaccurate. I may not understand fully how TimeBank data is calculated, but it seems to me like either TimeBank.start needs a value accurate to the minute, or, TimeBank.date_worked needs to fully replace TimeBank.start. 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/bushwickfc/bfc-members/issues/16#issuecomment-127042389


-- 
Mike Hogg <me@themikehogg.com>

**Comment by stratigos** [comment]
@themikehogg No worries, and thanks for the detailed advice. As for placing the `$().datetimepicker();` call inside the (time_banks.js) `$(document).ready()` call, I found it was still being called before bootstrap.js finished loading, on initial page load. So basically the first time a user arrived on the time_banks or members/time_banks pages, the datepicker wasnt responding to input. Refreshing the page would allow the datetimepicker to respond to mouse clicks. Im not sure why, I had thought it was due to turbolinks, but Im really not sure.

When I was debugging with firebug, the network graph and JS console would show that the time_banks.js was being called before bootstrap.js was finished downloading, on initial page load. On page refresh, no JS assets were downloaded as they were all cached, and time_banks.js doc.ready() was called after bootstrap.js was finished. It seems to me like this problem would be avoided by putting boostrap into the asset pipeline, just below jquery. 

I will try to download the datetimepicker.js and moment.js resources and place them in the (vendor) assets directories as is stated in [that SO post](http://stackoverflow.com/questions/18371318/installing-bootstrap-3-on-rails-app). However, if the call to `$().datetimepicker();` happens before bootstrap.js loads, it may run into the same exact problem.

**Comment by stratigos** [comment]
Ok so the culprit here is Turbolinks. Ive done quite a bit of research on the subject and I must agree with what appears to be the general consensus amongst the Rails community: Tubolinks serves a very specific purpose, and tends to cause more problems than it solves when simply used by default on every Rails application. 

However, there is a simple solution. We should call `$(document).on('page:load', SOMEFUNC);` for initial navigation to each page, and `$(document).ready(SOMEFUNC);` for any subsequent refresh. Both are essentially necessary to work with Turbolinks on, and the DataTables plugin being used. Solution derived from [here](http://stackoverflow.com/questions/18770517/rails-4-how-to-use-document-ready-with-turbo-links), though oddly Im actually using this technique on a JS heavy Rails site I made a few months ago, and just completely forgot about it.

Its not perfect because you have to bind twice, but it does beat the other solutions I found (discussed below).

--------

Since Turbolinks just performs an AJAX refresh of page content, and updates the `<title>` element, and the rest of `<head>` is not loaded, the `$(document).ready()` calls are never called, and the nodes on which events are bound are lost. This explains why my JS in the erb view template worked fine, but my JS within an assets/javascripts/*.js file did nothing on initial page load. It also explains why everything worked as expected upon page refresh. One could craft the application's JS to work around this pattern, but one should also ask if Turbolinks is really the right choice for the application in the first place.

So the [best](http://stackoverflow.com/questions/17881384/jquery-gets-loaded-only-on-page-refresh-in-rails-4-application) [solution](http://stackoverflow.com/questions/17386740/rails-4-turbolinks-and-jquery-dynamic-links-not-playing-nice) Ive [found](http://stackoverflow.com/questions/19043134/making-jquery-works-with-turbolinks) is to use the jquery.turbolinks gem. The datetimepicker works as expected using the [jquery.turbolinks](https://github.com/kossnocorp/jquery.turbolinks) gem! Unfortunately, this raises a huge problem with [DataTables](http://www.datatables.net/), namely, a very obtrusive alert popup appears every time you load any index page with a table on it. This bug, causing DataTables to alert with the following error, makes it unviable for our use:

`DataTables warning: table id=DataTables_Table_0 - Cannot reinitialise DataTable. For more information about this error, please see http://datatables.net/tn/3`

Please let me know if there is some trivial thing about DataTables that I'm missing that allows it to work here. 

I believe the ideal solution is just remove Turbolinks. I would also prefer to use `jquery.turbolinks` as a solution but the DataTables alert error makes for a wretched UX. I certainly frown upon mixing the JS within the erb view templates, and dont know any other solutions. For now, Ill just bind to `doc.ready()` and `doc.on(page:load)`. 

--------

Just for reference, here are some blog [posts](http://blog.joshsoftware.com/2013/06/28/troubleshooting-turbolinks-with-rails-4/) I [read](http://blog.flightswithfriends.com/post/53943440505/how-to-disable-turbolinks-in-rails-4) [about](https://blog.engineyard.com/2015/a-look-at-rails-5) [turbolinks](https://www.amberbit.com/blog/2015/4/22/why-rails-5-turbolinks-3-action-cable-matter-and-why-dhh-was-right-all-along/). Also its important to note that the Rails team is creating a Turbolinks v.3 for Rails 5, and it wont be included by default anymore. At this time Im not exactly opinionated, as I'm just trying to find the best conventions to follow (concerning Rails). If there is a better argument for using Turbolinks, that also allows for a solution to conventional use of doc.ready(), that doesnt break the DataTables feature, I am all ears.

**Comment by stratigos** [comment]
Hey also, do you guys feel there should be a migration to update time_banks.date_worked to the values from time_banks.start?  Also, perhaps updating time_banks.hours_worked to the difference of start or finished (depending on time_type)?

Right now it will just be lots of zeroes for date_worked / hours_worked, which might confuse some users.

**Comment by themikehogg** [comment]
Yes, that makes sense to me.


On Fri, 07 Aug 2015 07:40:34 -0700
Todd <notifications@github.com> wrote:

> Hey also, do you guys feel there should be a migration to update time_banks.date_worked to the values from time_banks.start?  Also, perhaps updating time_banks.hours_worked to the difference of start or finished (depending on time_type)?
> 
> Right now it will just be lots of zeroes for date_worked / hours_worked, which might confuse some users. 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/bushwickfc/bfc-members/issues/16#issuecomment-128717016


-- 
Mike Hogg <me@themikehogg.com>

**Comment by flah00** [comment]
I didn't get a chance to review what was happening sooner, sorry for that.
Please check out my branch, to see the virtual methods in action, it's a much less risky change.
pc-change-start-finish

**Comment by stratigos** [comment]
Ah ok, very cool, lets discuss in the meeting today if there is time :)

**Comment by stratigos** [comment]
hey @flah00 and @themikehogg Im wondering if you or anyone else would have some advice on testing virtual attributes.

It seems to me that, within the models/time_bank.rb tests, anytime I instantiate a new TimeBank, I cant read or write the `date_worked` or `hours_worked` attribute. Ive tried setting it via `TimeBank.new({attrs})`, and a number of other ways. The `hours_worked` attribute will always return the `rescue` block from the `hours()` callback (from class TimeBank), and essentially be assigned as `0.0`. 

I have this mostly wrapped up on branch pc-change-start-finish, I just cant seem to get any tests working concerning the `date_worked`, `hours_worked` attributes. The setter methods dont seem to do anything within the test suite. I'm also unsure how to test the `date_hack()` method of TimeBank. Any ideas on how to get test coverage for the virtual attributes?

**Comment by themikehogg** [comment]
Todd,
I think this was where I got stuck, although I was just trying to hook up the html element to the virtual attribute.  I opened this today, and spent my hour getting context back and trying to figure out how to seed my test db.  Sorry I wasn't ablet to get to running tests yet.  I will try another hour on this next weekend.  

I do note that you said you created another virtual attribute for date_worked and I was just using start_date for that, and changing the label. Ultimately not going to solve our problem, but it's less change if it can be done that way.

Also, is your work committed to your branch? Then I should be able to see how you set up your virtuals.

Mike

On Thu, 13 Aug 2015 15:11:08 -0700
Todd <notifications@github.com> wrote:

> hey @flah00 and @themikehogg Im wondering if you or anyone else would have some advice on testing virtual attributes.
> 
> It seems to me that, within the models/time_bank.rb tests, anytime I instantiate a new TimeBank, I cant read or write the `date_worked` or `hours_worked` attribute. Ive tried setting it via `TimeBank.new({attrs})`, and a number of other ways. The `hours_worked` attribute will always return the `rescue` block from the `hours()` callback (from class TimeBank), and essentially be assigned as `0.0`. 
> 
> I have this mostly wrapped up on branch pc-change-start-finish, I just cant seem to get any tests working concerning the `date_worked`, `hours_worked` attributes. The setter methods dont seem to do anything within the test suite. I'm also unsure how to test the `date_hack()` method of TimeBank. Any ideas on how to get test coverage for the virtual attributes?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/bushwickfc/bfc-members/issues/16#issuecomment-130863370


-- 
Mike Hogg <me@themikehogg.com>

**Comment by stratigos** [comment]
@themikehogg I am assuming you are replying to a comment still left behind in the email thread associated with/tracking this issue. I scoured stackoverflow all weekend after leaving that comment, then realized what the problem was (I was using the setter methods to assign a value to `timebank.attributes`, instead of creating instance vars for `@date_worked` and `@hours_worked`). Everything works as expected now, so I deleted that comment.

@flah00 had started the most recent branch for this issue, which is the branch used in the pull request above. Take a look at the changeset in #22 if you would like to see how @flah00 initially set up the virtual attributes. I agree that it is more clear with a separate virtual attr for `@date_worked` (in addition to `@hours_worked`).
